GabijaBern/assignment2/BinaryTreeAncestors.java

```diff
@@ -0,0 +1,251 @@
+/**
+*Class BinaryTreeAncestors is the solution for the first question in second assignment 
+*Here I have implemented my own tree where only unique values are available
+*/
+import java.util.Queue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+
+public class BinaryTreeAncestors
+{
+	public static void main(String[] args)
+	{
+		BTree<Integer> test = new BTree<Integer>();
+		test.treeRoot = new Node<Integer>(0);
+		test.insert(test.treeRoot, 1);
+		test.insert(test.treeRoot, 2);
+		test.insert(test.treeRoot, 3);
+		test.insert(test.treeRoot, 4);
+		test.insert(test.treeRoot, 5);
+		test.insert(test.treeRoot, 6);
+		
+		//test.findAncestors(test, new Node<Integer>(3));
+		
+		//Node<Integer> node = new Node<Integer>(null);
+		//node = test.findCommonAncestor(test, test.treeRoot.left.treeRoot.left.treeRoot, test.treeRoot.right.treeRoot);
+		//System.out.println(node);
+		//System.out.println(test.findCommonAncestor2(test, test.treeRoot.left.treeRoot.left.treeRoot, test.treeRoot.right.treeRoot.right.treeRoot));
+	}
+}
+/*
+ * Generic class for Nodes
+ * 
+ * Here we override toString that is used while testing the methods
+ */
+class Node<T> 
+{
+	public T value;
+	public BTree<T> left = null, right = null;
+	
+	public Node(T valueToSet)
+	{
+		this.value = valueToSet;
+	}
+	
+	@Override
+	public String toString()
+	{
+		return "" + value;
+	}
+}
+/*
+ * Binary tree implementation with insertion and ancestors methods
+ */
+class BTree<T>

```

Comment:
Having class both Node and TreeRoot makes your code quite long and verbose. Did you consider just using Node? If you chose not to, why did you make this decision?

Reply:
This is just the design decision that I have been taught in java classes. Our teacher always argued that having two separate classes is more real world implementation, but I do agree that it makes it quite verbose.

Reply:
OK. So the most real world decision is start with the simplest design that works. Once you have the need for extra complexity you add it at that point. It is a process called "refactoring", you can read a lot about it at https://refactoring.com/ if you are interested, which has a link to a good book.

There is a design principle called YAGNI (You ain't gonna need it) which you might be interested in reading about here: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it